internals/src/hex/buf_encoder.rs

```diff
@@ -8,10 +8,32 @@ pub use out_bytes::OutBytes;
 
 use super::Case;
 
+/// Trait for types that can be soundly converted to `OutBytes`.
+///
+/// To protect the API from future breakage this sealed trait guards which types can be used with
+/// the `Encoder`. Currently it is implemented for byte arrays of various interesting lengths.
+///
+/// ## Safety
+///
+/// This is not `unsafe` yet but the `as_out_bytes` should always return the same reference if the
+/// same reference is supplied. IOW the returned memory address and length should be the same if
+/// the input memory address and length are the same.
+///
+/// If the trait ever becomes `unsafe` this will be required for soundness.
+pub trait AsOutBytes: out_bytes::Sealed {
+    /// Performs the conversion.
+    fn as_out_bytes(&self) -> &OutBytes;
+
+    /// Performs the conversion.
+    fn as_mut_out_bytes(&mut self) -> &mut OutBytes;
```

Comment:
Note one would normally have two traits - one for `mut` the other for non-`mut`. The limitation is that immutable version of `AsOutBytes` can not be implemented for `&T where T: AsOutBytes`.

I think this is fine, we really don't want to deal with immutable slices beyond providing non-weird API of `BufEncoder`